### PR TITLE
Added support for MPS (Apple Silicon)

### PR DESCRIPTION
**Pull request by fnachon**
A few code fixes to make RFD3, RF3, and MNPNN run on MacOS with M1-M5 chips.
Mostly handling of Bfloat16 -> Float32 for mps.

### CODE REVIEW

File: models/rfd3/src/rfd3/model/layers/block_utils.py [hunk 1]
```diff
@@ -137,9 +158,9 @@ def scatter_add_pair_features(P_LK_tgt, P_LK_indices, P_LA_src, P_LA_indices):
     # Handle case when indices and P_LA don't have batch dimensions
     B, L, k = P_LK_indices.shape
     if P_LA_indices.ndim == 2:
-        P_LA_indices = P_LA_indices.unsqueeze(0).expand(B, -1, -1)
```
Comment (Ubiquinone-dot) [inline models/rfd3/src/rfd3/model/layers/block_utils.py:140]: are these contiguous calls necessary?

File: models/rfd3/src/rfd3/model/layers/block_utils.py [hunk 2]
```diff
@@ -407,21 +428,23 @@ def extend_index_mask_with_neighbours(
     inf = torch.tensor(float("inf"), dtype=D_LL.dtype, device=device)
 
     # 1. Selection of sequence neighbours
-    all_idx_row = torch.arange(L, device=device).expand(L, L)
-    indices = torch.where(mask, all_idx_row, inf)  # sentinel inf if not-forced
+    # Use .repeat() instead of .expand() to produce a contiguous tensor — MPS does
+    # not handle non-contiguous inputs to torch.where correctly.
```
Comment (Ubiquinone-dot) [inline models/rfd3/src/rfd3/model/layers/block_utils.py:432]: make this an if mps else do the old version; old version is faster afaik

**Review by Ubiquinone-dot — COMMENTED** [review]
generally close to being ready, just a few comments for clarity

**Review by Ubiquinone-dot — COMMENTED** [review]
## Review of MPS Support PR

### Overall

Good PR — the changes are well-scoped and the MPS fallback patterns are sound. A few items to address:

### 1. `block_utils.py` line 433 — `.repeat()` allocates unnecessarily on non-MPS

The unconditional `.repeat(L, 1)` replaces a zero-copy `.expand(L, L)` with an O(L²) allocation on **every** backend. This should be guarded with an if-else:

```python
# MPS does not handle non-contiguous inputs to torch.where correctly,
# so use .repeat() (allocates) there; .expand() (zero-copy view) elsewhere.
if device.type == "mps":
    all_idx_row = torch.arange(L, device=device).unsqueeze(0).repeat(L, 1)
else:
    all_idx_row = torch.arange(L, device=device).unsqueeze(0).expand(L, L)
```

### 2. `.contiguous()` calls throughout `block_utils.py` — safe, keep them

The added `.contiguous()` calls (lines 161, 163, 178, 179, 193, 194, 374, 446, 447) are all no-ops on already-contiguous tensors (CUDA/CPU expand often returns contiguous tensors for these shapes). No performance concern — these are fine to keep unconditionally.

### 3. `get_sparse_attention_indices` line 463 — duplicate assertion skip

```python
if indices.device.type != "mps" and (indices[..., 1:] == indices[..., :-1]).any():
```

This silently skips a correctness assertion on MPS. Is this due to actual false positives from MPS numerical quirks in `topk`/`sort`, or is it masking a real issue? Consider at minimum logging a warning on MPS instead of silently skipping:

```python
has_dupes = (indices[..., 1:] == indices[..., :-1]).any()
if has_dupes:
    if indices.device.type == "mps":
        logger.warning("Duplicate sparse attention indices detected on MPS (known quirk)")
    else:
        raise AssertionError("Tensor has duplicate elements along the last dimension.")
```

### 4. `scatter_mean` utility — looks correct

The MPS fallback via `scatter_add + count` is correct and efficient (single-slice count trick avoids full allocation). The non-MPS path correctly delegates to `index_reduce`. The removal of `.clone()` from the original call sites is also correct since `index_reduce` returns a new tensor.

### 5. `frames.py` and `alignment.py` — `det` intermediate variable

Good catch. Storing `torch.linalg.det(R)` in a named variable before passing to `torch.sign()` avoids MPS issues with in-place ops on autograd graph leaves. Safe on all backends.

### 6. Ruff version mismatch

The PR author likely formatted with a different ruff version than the pinned `0.8.3`. Three files need reformatting: `rf3/pairformer_layers.py`, `rfd3/pairformer_layers.py`, and `foundry/utils/torch.py`. Run `make format` with `ruff==0.8.3` to fix.

**Comment by Ubiquinone-dot** [comment]
Superseded by #260 which includes these changes with additional fixes (expand/repeat if-else for backward compat on non-MPS backends, ruff 0.8.3 formatting for CI).

**Comment by Ubiquinone-dot** [comment]
Have merged this now via PR #257 since i couldn't push to this branch!
